File: project/src/nwr/gui/controls/EditBox.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package nwr.gui.controls;

import jzrlib.core.Rect;
import jzrlib.utils.TextUtils;
import nwr.engine.ControlStyles;
import nwr.engine.KeyEventArgs;
import nwr.engine.KeyPressEventArgs;
import nwr.engine.MouseEventArgs;
import nwr.engine.BaseControl;
import nwr.engine.BaseScreen;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class EditBox extends BaseControl
{
    private int fCaretPos;
    private String fText;

    @Override
    protected void doKeyDownEvent(KeyEventArgs eventArgs)
    {
        switch (eventArgs.Key) {
            case GK_BACK:
                if (this.fCaretPos > 0) {
                    this.fText = this.fText.substring(0, this.fCaretPos - 1) + this.fText.substring(this.fCaretPos);
                    this.fCaretPos--;
                }
                break;

            case GK_DELETE:
                if (this.fCaretPos >= 0 && this.fCaretPos < this.fText.length()) {
                    this.fText = this.fText.substring(0, this.fCaretPos) + this.fText.substring(this.fCaretPos + 1);
                }
                break;

            case GK_RIGHT:
                if (this.fCaretPos < this.fText.length()) {
                    this.fCaretPos++;
                }
                break;

            case GK_LEFT:
                if (this.fCaretPos > 0) {
                    this.fCaretPos--;
                }
                break;
        }

        super.doKeyDownEvent(eventArgs);
    }

    @Override
    protected void doKeyPressEvent(KeyPressEventArgs eventArgs)
    {
        if (super.getFont().isValidChar(eventArgs.Key)) {
            this.fText = new StringBuilder(this.fText).insert(this.fCaretPos, eventArgs.Key).toString();
            this.fCaretPos++;
        }
        super.doKeyPressEvent(eventArgs);
    }

    @Override
    protected void doMouseDownEvent(MouseEventArgs eventArgs)
    {
        super.doMouseDownEvent(eventArgs);

        eventArgs.X -= (super.getHeight() - super.getFont().Height) / 2;
        int sh = 0;
        String ln = "";

        int num = (this.fText != null) ? this.fText.length() : 0;
        for (int i = 1; i <= num; i++) {
            ln += this.fText.charAt(i - 1);

            int nw = super.getFont().getTextWidth(ln);
            if (eventArgs.X >= sh && eventArgs.X <= nw) {
                this.fCaretPos = i;
                break;
            }

            sh = nw;
        }
    }

    @Override
    protected void doPaintEvent(BaseScreen screen)
    {
        Rect crt = super.getClientRect();
        if ((super.ControlStyle.contains(ControlStyles.csOpaque))) {
            screen.fillRect(crt, BaseScreen.clBlack);
        }
        CtlCommon.drawCtlBorder(screen, crt);

        screen.setTextColor(BaseScreen.clGold, true);

        int th = super.getFont().Height;
        int mg = (super.getHeight() - th) / 2;
        int x = crt.Left + mg;
        int y = crt.Top + mg;
        screen.drawText(x, y, this.fText, 0);

        int tw;
        if (this.fCaretPos == 0) {
            tw = 0;
        } else {
            tw = screen.getTextWidth(this.fText.substring(0, this.fCaretPos)) + 1;
        }

        int cw;
        if (this.fCaretPos >= this.fText.length()) {
            cw = 8;
        } else {
            cw = screen.getTextWidth(this.fText.substring(this.fCaretPos - 1, this.fCaretPos + 1)); // FIXME: need caret line
        }
        x += tw;
        y = y + th + 1;
        screen.drawLine(x, y, x + cw, y, BaseScreen.clGold);

        //screen.drawText(x + tw, y + 3, "_", 0); // cursor
    }

    @Override
    public Rect getIntRect()
    {
        return new Rect(2, 2, super.getWidth() - 3, super.getHeight() - 3);
    }

    public String getText()
    {
        return this.fText;
    }

    public void setText(String value)
    {
        if (!TextUtils.equals(this.fText, value)) {
            this.fText = value;

            if (TextUtils.equals(this.fText, "")) {
                this.fCaretPos = 0;
            } else {
                this.fCaretPos = this.fText.length();
            }
        }
    }

    public EditBox(BaseControl owner)
    {
        super(owner);
        this.setText("");

        if (super.getFont() != null) {
            super.setHeight(super.getFont().Height + 8);
        } else {
            super.setHeight(20);
        }

        super.setWidth(200);
    }
}

File: project/GUI/Controls/EditBox.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  this file is part of "NorseWorld: Ragnarok".
 *
 *  this program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  this program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System.Text;
using BSLib;
using ZRLib.Engine;

namespace NWR.GUI.Controls
{
    public sealed class EditBox : BaseControl
    {
        private int fCaretPos;
        private string fText;

        protected override void DoKeyDownEvent(KeyEventArgs eventArgs)
        {
            switch (eventArgs.Key) {
                case Keys.GK_BACK:
                    if (fCaretPos > 0) {
                        fText = fText.Substring(0, fCaretPos - 1) + fText.Substring(fCaretPos);
                        fCaretPos--;
                    }
                    break;

                case Keys.GK_DELETE:
                    if (fCaretPos >= 0 && fCaretPos < fText.Length) {
                        fText = fText.Substring(0, fCaretPos) + fText.Substring(fCaretPos + 1);
                    }
                    break;

                case Keys.GK_RIGHT:
                    if (fCaretPos < fText.Length) {
                        fCaretPos++;
                    }
                    break;

                case Keys.GK_LEFT:
                    if (fCaretPos > 0) {
                        fCaretPos--;
                    }
                    break;
            }

            base.DoKeyDownEvent(eventArgs);
        }

        protected override void DoKeyPressEvent(KeyPressEventArgs eventArgs)
        {
            if (Font.IsValidChar(eventArgs.Key)) {
                fText = (new StringBuilder(fText)).Insert(fCaretPos, eventArgs.Key).ToString();
                fCaretPos++;
            }
            base.DoKeyPressEvent(eventArgs);
        }

        protected override void DoMouseDownEvent(MouseEventArgs eventArgs)
        {
            base.DoMouseDownEvent(eventArgs);

            eventArgs.X -= (Height - Font.Height) / 2;
            int sh = 0;
            string ln = "";

            int num = (fText != null) ? fText.Length : 0;
            for (int i = 1; i <= num; i++) {
                ln += fText[i - 1];

                int nw = Font.GetTextWidth(ln);
                if (eventArgs.X >= sh && eventArgs.X <= nw) {
                    fCaretPos = i;
                    break;
                }

                sh = nw;
            }
        }

        protected override void DoPaintEvent(BaseScreen screen)
        {
            ExtRect crt = ClientRect;
            if ((ControlStyle.Contains(ControlStyles.сsOpaque))) {
                screen.FillRect(crt, BaseScreen.clBlack);
            }
            CtlCommon.DrawCtlBorder(screen, crt);

            screen.SetTextColor(BaseScreen.clGold, true);

            int th = Font.Height;
            int mg = (Height - th) / 2;
            int x = crt.Left + mg;
            int y = crt.Top + mg;
            screen.DrawText(x, y, fText, 0);

            int tw;
            if (fCaretPos == 0) {
                tw = 0;
            } else {
                tw = screen.GetTextWidth(fText.Substring(0, fCaretPos)) + 1;
            }

            int cw;
            if (fCaretPos >= fText.Length) {
                cw = 8;
            } else {
                cw = screen.GetTextWidth("" + fText[fCaretPos]);
            }
            x += tw;
            y = y + th + 1;
            screen.DrawLine(x, y, x + cw, y, BaseScreen.clGold);

            //screen.drawText(x + tw, y + 3, "_", 0); // cursor
        }

        public override ExtRect IntRect
        {
            get {
                return ExtRect.Create(2, 2, Width - 3, Height - 3);
            }
        }

        public string Text
        {
            get {
                return fText;
            }
            set {
                if ((fText != value)) {
                    fText = value;
    
                    if ((fText == "")) {
                        fCaretPos = 0;
                    } else {
                        fCaretPos = fText.Length;
                    }
                }
            }
        }

        public EditBox(BaseControl owner)
            : base(owner)
        {
            Text = "";

            if (Font != null) {
                Height = Font.Height + 8;
            } else {
                Height = 20;
            }

            Width = 200;
        }
    }
}
